TulparAudio/internal/source/SourceCollection.cpp

```diff
@@ -0,0 +1,1097 @@
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+float SourceCollection::GetSourcePlaybackProgress(SourceHandle source) const
+{
+    float result = 0.0f;
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint sampleOffset;
+    alGetSourcei(static_cast<ALuint>(source), AL_SAMPLE_OFFSET, &sampleOffset);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR == alErr)
+    {
+        audio::Source::State const state = GetSourceState(source);
+
+        if (sampleOffset > 0
+            && ((audio::Source::State::Playing == state)
+                || (audio::Source::State::Paused == state))
+        )
+        {
+            result = static_cast<float>(sampleOffset) / static_cast<float>(m_sourceMeta.at(source).activeSampleCount);
+        }
+    }
+    else
+    {
+        LOG_AUDIO->Warning("Source #{}: get playback progress: {:#x}", source, alErr);
+    }
+
+    return result;
+}
+
+bool SourceCollection::SetSourcePlaybackProgress(SourceHandle source, float value)
+{
+    LOG_AUDIO->Debug("Source #{}: set playback progress {}%", source, value);
+
+    audio::Buffer buffer = GetSourceActiveBuffer(source);
+
+    ALint sampleOffset = static_cast<ALint>(std::round(static_cast<float>(m_sourceMeta[source].activeSampleCount) * value));
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    alSourcei(static_cast<ALuint>(source), AL_SAMPLE_OFFSET, sampleOffset);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR != alErr)
+    {
+        LOG_AUDIO->Warning("Source #{}: set playback progress: {:#x}", source, alErr);
+    }
+
+    return AL_NO_ERROR == alErr;
+}
+
+audio::Source::State SourceCollection::GetSourceState(SourceHandle source) const
+{
+    audio::Source::State state = audio::Source::State::Unknown;
+
+    // clear error state
+    ALenum alErr = alGetError();
+
+    ALint alState;
+    alGetSourcei(static_cast<ALuint>(source), AL_SOURCE_STATE, &alState);
+
+    alErr = alGetError();
+
+    if (AL_NO_ERROR == alErr)
+    {
+        using audio::Source;
+
+        switch (alState)
+        {
+            case AL_INITIAL:
+            {
+                state = Source::State::Initial;
+                break;
+            }
+            case AL_PLAYING:
+            {
+                state = Source::State::Playing;
+                break;
+            }
+            case AL_PAUSED:
+            {
+                state = Source::State::Paused;
+                break;
+            }
+            case AL_STOPPED:
+            {
+                state = Source::State::Stopped;
+                break;
+            }
+            default:
+            {
+                state = Source::State::Unknown;
+                break;
+            }
+        }
+    }
+    else
+    {
+        LOG_AUDIO->Warning("Source #{}: get state: {:#x}", source, alErr);
+    }
+
+    return state;
+}
+
+audio::Source::Type SourceCollection::GetSourceType(SourceHandle source) const
```

Comment:
Shouldn't it be static?

Reply:
Answered in https://github.com/Godlike/Tulpar/pull/6#discussion_r244268061